crates/flipperzero/src/furi/log/metadata.rs

```diff
@@ -0,0 +1,535 @@
+    /// *disabled*, but **should not** be used for determining if something is
+    /// *enabled*.
+    ///
+    /// [`Level`]: super::Level
+    #[inline(always)]
+    pub fn current() -> Self {
+        match unsafe { sys::furi_log_get_level() } {
+            // Default log level is defined in `furi/core/log.c` in the FlipperZero firmware.
+            sys::FuriLogLevel_FuriLogLevelDefault => Self::INFO,
+            sys::FuriLogLevel_FuriLogLevelNone => Self::OFF,
+            sys::FuriLogLevel_FuriLogLevelError => Self::ERROR,
+            sys::FuriLogLevel_FuriLogLevelWarn => Self::WARN,
+            sys::FuriLogLevel_FuriLogLevelInfo => Self::INFO,
+            sys::FuriLogLevel_FuriLogLevelDebug => Self::DEBUG,
+            sys::FuriLogLevel_FuriLogLevelTrace => Self::TRACE,
+            #[cfg(debug_assertions)]
+            unknown => unreachable!(
+                "/!\\ `LevelFilter` representation seems to have changed! /!\\ \n\
+                This is a bug (and it's pretty bad). Please contact the `flipperzero` \
+                maintainers. Thank you and I'm sorry.\n \
+                The offending repr was: {:?}",
+                unknown,
+            ),
+            #[cfg(not(debug_assertions))]
+            _ => unsafe {
+                // Using `unreachable_unchecked` here (rather than
+                // `unreachable!()`) is necessary to ensure that rustc generates
+                // an identity conversion from integer -> discriminant, rather
+                // than generating a lookup table. We want to ensure this
+                // function is a single `bl` instruction (sometimes followed by
+                // a `subs` instruction to handle `FuriLogLevelDefault`) if at
+                // all possible, because it is called *every* time a logging
+                // callsite is hit; and it is (potentially) the only code in the
+                // hottest path for skipping a majority of callsites when level
+                // filtering is in use.
+                //
+                // safety: This branch is only truly unreachable if we guarantee
+                // that no values other than the possible enum discriminants
+                // will *ever* be present. The log filter is initialized by the
+                // Flipper Zero SDK to `FuriLogLevelDefault`, which is not a
+                // valid `LevelFilter` discriminant but is specifically handled
+                // above. It is set either internally by the Flipper Zero, or
+                // through the Flipper Zero SDK. The latter we expose via the
+                // `set_max` function, which takes a `LevelFilter` parameter;
+                // this restricts the inputs to `set_max` to the set of valid
+                // discriminants. Therefore, **as long as `furi_log_set_level`
+                // is only ever called by `set_max`**, this is safe.
+                core::hint::unreachable_unchecked()
+            },
```

Comment:
For context, without this `unreachable_unchecked()` (instead relying on `unreachable!()`, and removing its useful error message entirely), `hello-rust.fap` (with its single `info!()` usage) increases from 1080 bytes to 2920 bytes.

Reply:
Oh nice. I didn't know about `unreachable_unchecked()`. I ended up resorting to a rather hacky solution.

https://github.com/flipperzero-rs/flipperzero/blob/dfb671366e1e19f99e0ed85ac8491fb06bc8bd8a/crates/sys/src/lib.rs#L33